app/src/main/java/org/stepik/android/presentation/certificate/CertificatesPresenter.kt

```diff
@@ -134,6 +142,50 @@ constructor(
             )
     }
 
+    fun updateCertificate(certificate: Certificate) {
+        val oldState = state
+
+        val currentItems = (oldState as? CertificatesView.State.CertificatesRemote)?.certificates
+            ?: (oldState as? CertificatesView.State.CertificatesCache)?.certificates
+            ?: return
+
+        isBlockingLoading = true
+        compositeDisposable += certificatesInteractor
+            .updateCertificate(certificate)
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .doFinally { isBlockingLoading = false }
+            .subscribeBy(
+                onSuccess = { updatedCertificate ->
+                    val updatedItems = currentItems.transform {
+                        map { certificateViewItem ->
+                            if (certificateViewItem.certificate.id == updatedCertificate.id) {
+                                certificateViewItem.copy(certificate = updatedCertificate)
+                            } else {
+                                certificateViewItem
+                            }
+                        }
+                    }
+
+                    state =
+                        when (oldState) {
+                            is CertificatesView.State.CertificatesRemote ->
+                                CertificatesView.State.CertificatesRemote(updatedItems)
+
+                            is CertificatesView.State.CertificatesCache ->
+                                CertificatesView.State.CertificatesCache(updatedItems)
+
+                            else ->
+                                oldState
+                        }
+                    view?.showChangeNameSnackbar(isSuccess = true)
+                },
+                onError = {
+                    view?.showChangeNameSnackbar(isSuccess = false)
```

Comment:
Will we lose user input In case of error? It will be better to use approach from Compose Comment dialog or new course review.